design/014-secretstore-generator-v2.md

```diff
@@ -0,0 +1,376 @@
+  GetSecret(SecretStoreSpec, ExternalSecretDataRemoteRef) ([]byte, error)
+  PushSecret(SecretStoreSpec, *corev1.Secret, PushSecretData) error
+  DeleteSecret(SecretStoreSpec, PushSecretRemoteRef) error
+  SecretExists(SecretStoreSpec, PushSecretRemoteRef) (bool, error)
+  GetAllSecrets(SecretStoreSpec, ExternalSecretFind) (map[string][]byte, error)
+  Validate(SecretStoreSpec) (admission.Warnings, error)
+  Capabilities(SecretStoreSpec) SecretStoreCapabilities
+}
+
+type GeneratorV2 interface {
+  Generate(GeneratorSpec) (map[string][]byte, GeneratorProviderState, error)
+  Cleanup(GeneratorSpec, GeneratorProviderState) error
+}
+```
+
+### Out-of-Tree Providers Maintenance
+
+#### Deployment
+
+Out-of-tree providers are separate projects with their own repos, images, and Helm charts. Users deploy ESO and the providers they need. ESO connects to providers via a Kubernetes Service indicated by `providerConfig.address`. Co-locating providers as sidecars is **discouraged** to preserve isolation and scalability.
+
+#### Governance
+
+- One repo per "official" eso maintained provider (e.g., `provider-aws`).
+- Promotion lifecycle (experimental → stable).
+- CODEOWNERS and standard PR workflows per provider.
+- A collective community repo `provider-contrib` for community-maintained providers.
+
+### Notes/Constraints/Caveats
+
+- Do not implement Unix domain sockets for sidecars; providers should run as independent deployments to ensure horizontal scalability, separate network policies, and stronger isolation.
+- Cluster-scoped resources delegate to namespaced providers; referent authentication is explicit via `authenticationScope`.
+
+### Risks and Mitigations
+
+- Operational overhead: Users manage separate provider deployments 
+    - mitigated by dedicated Helm charts and independent versioning per provider.
+- Maintenance overhead: every provider needs similar infrastructure such as a helm chart, e2e test framework and test cases, a common library to bootstrap the provider's GRPC server, metrics, initialising clients etc.
+    - mitigated by providing one or more `common` repos, e.g. `provider-common` which host the shared code among eso-core and all providers (similar to https://github.com/fluxcd/pkg).
+- TLS management: someone needs to manage the TLS certs/keys: do we push it to the user or do we provide it ourselves?
+    - mitigated by (1) implement certificate management with `cert-controller` and  provide an integration mechanism with cert-manager.
+
+## Design Details
+
+### Test Plan
+
+This plan validates core changes in eso-core and out-of-tree providers, including security boundaries and migration flows.
+
+#### Prerequisite testing updates
+- Create a provider conformance test suite (library) that providers run against locally and in CI (covers GetSecret, GetAllSecrets, Push/Delete/Exists when implemented, error semantics, auth, TLS).
+- Add a fake gRPC provider for tests to simulate success, latency, failures, and version skew.
```

Comment:
in my experience this is not required, we simply need a gRPC server to start locally. Just my 2 cents though. Especially simulating success latency and so on can be done via handlers.